### PR TITLE
Refactor TableScanContext to avoid noisy updates while adding field

### PR DESCRIPTION
**Pull request by yittg**
When i am adding field into TableScanContext in #4115 , it would introduce methods updates for others unrelated fields.

Would like to refactor it to keep history clean while adding or removing fields.

cc @rdblue .

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/TableScanContext.java [hunk 1]
```diff
@@ -54,10 +54,18 @@
     this.toSnapshotId = null;
   }
 
-  private TableScanContext(Long snapshotId, Expression rowFilter, boolean ignoreResiduals,
-                           boolean caseSensitive, boolean colStats, Schema projectedSchema,
-                           Collection<String> selectedColumns, ImmutableMap<String, String> options,
-                           Long fromSnapshotId, Long toSnapshotId) {
+  private TableScanContext(
+      Long snapshotId,
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/TableScanContext.java:58]: While I personally like this format, in Iceberg we stick to the version that was here before with all the args grouped together and left justified with the first arg

**Comment by RussellSpitzer** [comment]
This feels like a good idea to me, although I'm hoping we don't have to make that many more additions to TableScanContext. Do you have a lot of fields you are thinking about adding?

File: core/src/main/java/org/apache/iceberg/TableScanContext.java [hunk 2]
```diff
@@ -131,37 +133,124 @@ Schema projectedSchema() {
...
 
   TableScanContext withOption(String property, String value) {
-    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
-    builder.putAll(options);
-    builder.put(property, value);
-    return new TableScanContext(snapshotId, rowFilter, ignoreResiduals,
-        caseSensitive, colStats, projectedSchema, selectedColumns, builder.build(), fromSnapshotId, toSnapshotId);
+    ImmutableMap.Builder<String, String> mapBuilder = ImmutableMap.builder();
+    mapBuilder.putAll(options);
+    mapBuilder.put(property, value);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableScanContext.java:146]: This rename looks unnecessary.

**Comment by rdblue** [comment]
We rarely change this class, so I'm not sure it's worth all the reformatting and extra builder, but I'm okay with this if @RussellSpitzer is in favor.

**Comment by RussellSpitzer** [comment]
> We rarely change this class, so I'm not sure it's worth all the reformatting and extra builder, but I'm okay with this if @RussellSpitzer is in favor.

I basically feel the same way, unless we had some plans to add a lot more to this class I think it's probably fine the way it is. That's why I asked above what the new planned fields were.

**Comment by yittg** [comment]
thanks @RussellSpitzer @rdblue .
I got touched this class in #4115 when I want to add an plan executor. and found the noisy update way. to avoid always thinking like ' let's refactor it NEXT time ' and leading to more and more meaningless updates, I tried to do this right now here.

If it's really rarely updated, it's definitely ok to leave it as the way it is and close this PR.

And given it's a simple change, i didn't open an issue in advance to discuss it. Sorry for that.

**Comment by yittg** [comment]
thanks @RussellSpitzer @rdblue .
I got touched this class in #4115 when I want to add an plan executor. and found the noisy update way. to avoid always thinking like ' let's refactor it NEXT time ' and leading to more and more meaningless updates, I tried to do this right now here.

If it's really rarely updated, it's definitely ok to leave it as the way it is and close this PR.

And given it's a simple change, i didn't open an issue in advance to discuss it. Sorry for that.

**Comment by yittg** [comment]
closing it.

**Pull request closed without merging**